extendedcopy.go

```diff
@@ -182,11 +179,31 @@ func findRoots(ctx context.Context, storage content.ReadOnlyGraphStorage, node o
 // For performance consideration, when using both FilterArtifactType and
 // FilterAnnotation, it's recommended to call FilterArtifactType first.
 func (opts *ExtendedCopyGraphOptions) FilterAnnotation(key string, regex *regexp.Regexp) {
+	filter := func(predecessors []ocispec.Descriptor, filtered []ocispec.Descriptor) []ocispec.Descriptor {
+		for _, p := range predecessors {
+			if value, ok := p.Annotations[key]; ok && (regex == nil || regex.MatchString(value)) {
+				filtered = append(filtered, p)
+			}
+		}
+		return filtered
+	}
+
 	fp := opts.FindPredecessors
 	opts.FindPredecessors = func(ctx context.Context, src content.ReadOnlyGraphStorage, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
 		var predecessors []ocispec.Descriptor
 		var err error
 		if fp == nil {
+			if rf, ok := src.(registry.ReferrerFinder); ok {
```

Comment:
Just a note and not related to this PR: With OCI Artifact support, a `registry.Repository` should always be a `registry.ReferrerFinder`. We will have `registry.Referrers(ctx context.Context, repo Repository) ([]string, error` later.